proto/zds.proto

```diff
@@ -0,0 +1,71 @@
+syntax = "proto3";
+
+// GRPC package - part of the URL. Service is added.
+// URL: /PACKAGE.SERVICE/METHOD
+package istio.workload.zds;
+
+option go_package="pkg/zdsapi";
+
+enum Version {
+  NOT_USED = 0;
+  V1 = 1;
+}
+
+message ZdsHello {
+  Version version = 1;
+}
+
+// Add a workload to the ztunnel. this will be accompanied by ancillary data contianing
+// the workload's netns file descriptor.
+message AddWorkload {
+  string uid = 1;
+}
+
+// Keep workload that we can't find in the fd cache. This can only be sent before SnapshotSent is sent
+// to signal ztunnel to not delete the workload if it has it.
+message KeepWorkload {
+  string uid = 1;
+}
+
+// Delete a workload from the ztunnel. Ztunnel should shutdown the workload's proxy.
+message DelWorkload {
+  string uid = 2;
+}
+
+// Let ztunnel know that a full snapshot was sent. Ztunnel should reconcile its internal state
+// and remove internal entries that were not sent.
+message SnapshotSent {
+}
+
+// Ztunnel ack message. If error is not empty, this is an error message.
+message Ack {
+  string error = 1;
+}
+
+/*
+Protocol details:
+on new connection from ztunnel to CNI, the CNI agent
+- will send all the existing payloads (that it has in its cache) to the ztunnel using AddWorkload message.
```

Comment:
Same as above - seems very complicated and for no clear reason.

Just send a json with all the data - how many pods are we talking about ? 200 ? One message in, one response out.

What can the CNI do in case of error - and what could be the error that requires this ping-pong ?

CNI is way overbloated and complex - I love the recent changes of using UDS and sending descriptors, but it can be done far simpler.

Reply:
This was done this way for simplicity.
The problem is not just the json size; its the combination of that with correlating the a pod uid in the AddWorkload message with the file descriptor.

The simplest way i could think of **making sure** we don't mess up the mapping is using PACKETSEQ socket type and send a single AddWorkload message and FD at the time.

Using a larger json will force us to use STREAM socket (instead of PACKETSEQ) and do more complex accounting of the workload stream and the FD stream...

Reply:
Is it not just a list of UIDs (json encoding, I suppose) and a list of FDs? Then correlating those (assume they are the same length or its an error)

Reply:
Conceptionally yes, it's list of uids and list of fds. But:

- If you use STREAM sockets you need to account for edge case where the list of uids and list of FDs come on different `recvmsg()` invocations. For example, `recvmsg` might return a shorter list of uids than the one sent (where the rest of the list will come in the next call to `recvmsg`).
- If we do it all in one message, we come pretty close to close to SCM_MAX_FD ( which is 253. GKE clusters can potentially go above this with a max of 256 pods per node). So we'll need to account for that and do batching in the code anyway..

Reply:
responding to @costinm [comment](https://github.com/istio/ztunnel/pull/747#issuecomment-1876093209) (i think it was meant as a reply here)

I think there is some confusion here, this protocol is not gRPC. It's a custom protocol between ztunnel (rust) and the node agent (go). 
Note:
- The reason protobuf was used, is so that we won't have to implement serialization code twice, in go and in rust. even if we want to use json for some reason, it would still be easier to use protobuf as the common schema.
- I think swagger/openapi schema is more complex than protobuf.
- Note that i have a [follow up](https://github.com/yuval-k/ztunnel/commit/4351650c7c773535aa13832940e0609465abc3a9#diff-f573b5300ba6d69c6a6be4ca2ca2c3cd7ec437ee459302c30f731969a8eb011b) where i plan to add more information to the messages
- The reason we wait for the reply is so that pod start-up is delayed until ztunnel opened its ports in the pod's netns (similar motivation to holdApplicationUntilProxyStarts).

Reply:
PACKETSEQ makes it easier as I know that the buffer was fully received. in a STREAM socket, there's a potential case that the json buffer and the FD stream will be split across different `recvmsg`.